### PR TITLE
Added BlindWalk potion for Google Code In 2016

### PR DESCRIPTION
**Pull request by arpitkamboj**
Added BlindWalk potion - Turns the player blind for 5 seconds along with speed decrease.

*Cervator edit: tied to GCI task https://codein.withgoogle.com/dashboard/task-instances/4855239574290432/*

### CODE REVIEW

File: assets/prefabs/potion/BlindWalkPotion.prefab [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    },
+    "Potion": {
+        "hasGenome": false,
+        "effects": [
+            {
+                "effect": "WALK",
+                "magnitude": 0.5,
+                "duration": 5100
+            },
+			{
```
Comment (rzats) [inline assets/prefabs/potion/BlindWalkPotion.prefab]: Also, the indentation here is a little odd - can you reformat the JSON file using IntelliJ?

**Review by rzats — CHANGES_REQUESTED** [review]
You'll actually have to implement the `NO_VISIBILITY` effect 🙂 For this you'll need to declare it in [`PotionCommonEffects`](https://github.com/Terasology/Potions/blob/master/src/main/java/org/terasology/potions/PotionCommonEffects.java), then implement in [`DrinkPotionAuthoritySystem`](https://github.com/Terasology/Potions/blob/master/src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java).

File: assets/prefabs/potion/BlindWalkPotion.prefab (hunk 1 above)
Comment (arpitkamboj) [reply assets/prefabs/potion/BlindWalkPotion.prefab]: Thanks for the advice... I was a little confused at first on how all the effects were implemented.... Now its all clear... Thanks. Will update soon..

**Comment by arpitkamboj** [comment]
Hey,
Even if I implement the 'No-Visibility' effect in common effects and potionauthoritysystem, there is no code to specify what the game has to do...

Please help

**Comment by rzats** [comment]
That's correct - actually implementing the code is also part of the task 🙂

**Comment by arpitkamboj** [comment]
Can you please give some idea on how to do that?

**Comment by rzats** [comment]
I've done some research on this and found out that making the effect work requires an API that's not yet merged (https://github.com/MovingBlocks/Terasology/pull/2449). Therefore, it's OK for you to submit a stub method/code block with a TODO note for later 🙂

**Comment by arpitkamboj** [comment]
Thanks.
I'll do that for now.
But I figured out that I need to create an entire new file for the No_visibility effect and set it to render a black screen for the given amount of time. I will work on this later.

-Arpit

File: assets/prefabs/potion/BlindWalkPotion.prefab [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    "Potion": {
+        "hasGenome": false,
+        "effects": [
+            {
+                "effect": "WALK",
+                "magnitude": 0.5,
+                "duration": 5100
+            },
+            {
+		"effect": "NO_VISIBILITY"
```
Comment (Cervator) [inline assets/prefabs/potion/BlindWalkPotion.prefab]: A few stray tab characters here, please replace with spaces :)

File: src/main/java/org/terasology/potions/PotionCommonEffects.java [hunk 3]
```diff
@@ -44,6 +44,7 @@
     public static final String FEATHER_FALL = "FEATHER_FALL";
     public static final String BRICK_FALL = "BRICK_FALL";
     public static final String MULTI_JUMP = "MULTI_JUMP";
+	public static final String NO_VISIBILITY = "NO_VISIBILITY";
```
Comment (Cervator) [inline src/main/java/org/terasology/potions/PotionCommonEffects.java]: Could you add something above this to indicate the pending effect implementation? Like:

`// TODO: Effect implementation awaiting API improvements, such as https://github.com/MovingBlocks/Terasology/pull/2449

File: src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java [hunk 4]
```diff
@@ -217,6 +217,9 @@ public void onPotionWithoutGenomeConsumed(DrinkPotionEvent event, EntityRef ref)
                 case PotionCommonEffects.EXPLOSIVE:
                     herbEffect = new ExplosiveEffect();
                     break;
+				case PotionCommonEffects.NO_VISIBILITY:
+                    herbEffect = new NoVisibilityEffect();
```
Comment (Cervator) [inline src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java:221]: Note that this class does not appear to be part of the PR. Forget to `git add` it? :-)

**Review by Cervator — CHANGES_REQUESTED** [review]
Heyo, few more tiny tweaks requested, although agreed on the effect implementation being problematic right at this moment so we can merge and approve in the meantime :-)

One additional note: the text on the potion icon will be *very* hard to see in-game (and even get reversed due to the way held items are rendered right now). Maybe just pick a color or simple icon? Maybe an outlined footprint (dashed line) instead

**Comment by Cervator** [comment]
And just to echo @rzats yeah it'll be fine to submit a more or less blank stub class for `NoVisibilityEffect` - we just need the code to actually compile before we can merge it :D

File: src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java (hunk 4 above)
Comment (arpitkamboj) [reply src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java:221]: OK I get it I need to add a file for NoVisibilityEffect to the potions/effects directory....

Thanks

**Comment by arpitkamboj** [comment]
Hey,
Please review the changes I made to the files.

Thanks
-Arpit

File: src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.potions.effect;
+
+import org.terasology.entitySystem.entity.EntityManager;
```
Comment (Cervator) [inline src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:18]: Most of these imports are unused, should show gray and underlined in IntelliJ - please remove those :-)

File: src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.logic.actions.*;
```
Comment (Cervator) [inline src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:21]: Wildcard imports are against our code style, although these will be removed anyway as unused. But if you configure IntelliJ right it'll avoid the wildcard imports / format them nicely in the future :-)

See http://stackoverflow.com/questions/3348816/intellij-never-use-wildcard-imports

File: src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import org.terasology.logic.health.*;
+import org.terasology.potions.HerbEffect;
+import org.terasology.potions.events.DrinkPotionEvent;
+
+/**
+ * Once the API for Custom Actions is implemented, the code for NoVisibility Effect goes here.
+ The effect will make the screen blank for the given amount of time.
+ */
+public class NoVisibilityEffect implements HerbEffect {
```
Comment (Cervator) [inline src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:30]: If you were compiling the code to make sure all was well you should've seen an error here. Like other effects even though this one is a placeholder pending API improvements anything implementing `HerbEffect` also must implement the `applyEffect` potion. Check out other effects to see how it should look - yours can probably be empty but needs to exist. It looks like you've got the Javadoc for it below, so maybe you just deleted the method header itself?

File: src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    /**
+     *
+     *
+     * @param instigator    The instigator who is applying this effect on the entity. It can be a herb, potion, etc.
+     * @param entity        The entity who the herb effect is being applied on.
+     * @param magnitude     The magnitude of the effect.
+     * @param duration      The duration of the effect in milliseconds.
+     
+	 Once the API for Custom Actions is implemented, the code for NoVisibility Effect goes here.
```
Comment (Cervator) [inline src/main/java/org/terasology/potions/effect/NoVisibilityEffect.java:39]: You have a couple tab characters here. Please replace with spaces :-)

File: src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java (hunk 4 above)
Comment (Cervator) [reply src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java:221]: You have the class now, but the DrinkPotionAuthoritySystem needs to import it before it'll be able to find it. Again something like IntelliJ should outright prompt you "Hey want to import this class from X?" :-) 

You also have tab characters again on the case statement line

File: README.md [hunk 9]
```diff
@@ -146,4 +151,6 @@ Send in PRs to this repository and feel free to add your name to the authors lis
 
 ## Credits for images:
 
-[Trekmarvel](https://github.com/Trekmarvel) for the base potion bottle graphics.
+* [Trekmarvel](https://github.com/Trekmarvel) for the base potion bottle graphics.
+* InvinciblePotionBottle.png, Regen2PotionBottle.png edited from base by Minege
+* MegaJumpBottle.png, GigaJumpBottle.png edited from base by Patrick Wang
```
Comment (Cervator) [inline README.md:156]: After this line you could put a quick note that "BlindWalkPotionBottle.png edited from base by digitalripperynr"

**Review by Cervator — CHANGES_REQUESTED** [review]
So at this point there's more small stuff that would be super easy to avoid if you were running out of IntelliJ and noticing its little recommendations. If you are at all able to get IntelliJ running with the source code showing in there that would be the best thing you could do right now :-)

Each issue is super easy to fix and I considered during it as part of the merge, but you really need to have a good setup to move forward with your coding adventure. Remember that IRC is available if you need help setting up!

**Comment by arpitkamboj** [comment]
Hey,
I apologise. I messed up everything.
Now installing IntelliJ.
I will close this PR and open a new one with all the changes (final this time).

Thanks
-Arpit

**Comment by arpitkamboj** [comment]
I am closing this PR and opening a new one...

**Pull request closed without merging**